src/ServerScriptService/InstanceComponents/Misc/Jetpack.lua

```diff
@@ -0,0 +1,220 @@
+local Players = game:GetService("Players")
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+local Trove = require(ReplicatedStorage.Packages.Trove)
+local Comm = require(ReplicatedStorage.Packages.Comm)
+local JetpackState = require(ReplicatedStorage.Common.Util.JetpackState)
+
+--[=[
+	@class Jetpack
+]=]
+local Jetpack = {}
+Jetpack.__index = Jetpack
+
+--[=[
+	Constructs a new Jetpack component.
+	@param instance Instance
+]=]
+function Jetpack.new(instance)
+	local serverComm = Comm.ServerComm.new(instance, "Jetpack")
+	local self = setmetatable({}, Jetpack)
+
+	self.Instance = instance
+	self._comm = serverComm
+	self._trove = Trove.new()
+
+	-- Set owning player
+	self._owner = Players:GetPlayerFromCharacter(instance.Parent)
+	instance.AncestryChanged:Connect(function()
+		self._owner = Players:GetPlayerFromCharacter(instance.Parent)
+	end)
+
+	-- Create state
+	self._state = JetpackState.new({
+		Boosting = instance:GetAttribute("Boosting");
+
+		Fuel = instance:GetAttribute("Fuel");
+		Capacity = instance:GetAttribute("Capacity");
+		FillRate = instance:GetAttribute("FillRate");
+		BurnRate = instance:GetAttribute("BurnRate");
+	})
+
+	-- Add state and comm to trove
+	self._trove:Add(self._state)
+	self._trove:Add(self._comm)
+
+	-- Subscribe to attribute changes
+	self._trove:Connect(instance:GetAttributeChangedSignal("Boosting"), function()
+		self._state:Dispatch("SetBoosting", instance:GetAttribute("Boosting"))
+	end)
+	self._trove:Connect(instance:GetAttributeChangedSignal("Fuel"), function()
+		self:SetFuel(instance:GetAttribute("Fuel"))
+	end)
+	self._trove:Connect(instance:GetAttributeChangedSignal("Capacity"), function()
+		self:SetCapacity(instance:GetAttribute("Capacity"))
+	end)
+	self._trove:Connect(instance:GetAttributeChangedSignal("FillRate"), function()
+		self:SetFillRate(instance:GetAttribute("FillRate"))
+	end)
+	self._trove:Connect(instance:GetAttributeChangedSignal("BurnRate"), function()
+		self:SetBurnRate(instance:GetAttribute("BurnRate"))
+	end)
+
+	-- Notify client about state events
+	self._trove:Add(self._state._silo:Subscribe(function(newState, oldState)
+		local function isNan(value)
+			return value ~= value
+		end
+
+		-- Set attributes for any values which have changed
+		for index, value in pairs(newState) do
+			local oldValue = oldState[index]
+			if value ~= oldValue or isNan(value) ~= isNan(oldValue) then
+				instance:SetAttribute(index, value)
+			end
+		end
+	end))
```

Comment:
This code is reaching across the object hierarchy as well - It should probably be rethought a bit.